source/common/common/packed_struct.h

```diff
@@ -0,0 +1,152 @@
+#pragma once
+
+#include <limits>
+#include <type_traits>
+
+#include "source/common/common/assert.h"
+#include "source/common/common/utility.h"
+#include "source/common/protobuf/utility.h"
+
+#include "absl/strings/str_cat.h"
+
+namespace Envoy {
+
+/**
+ * Helper class to pack elements of the same type into a struct like
+ * object, with the expectation that some or most of the elements are not set.
+ *
+ * The overhead of using this class as opposed to declaring the elements
+ * individually in another class or struct is 8 + (n+1) unaligned bytes, where n is
+ * the maximum number of elements in the struct.
+ *
+ * The way to use this class is to first define an enum class with the element
+ * names, and providing the enum class as the third template argument to
+ * PackedStruct, the first two being the type of element and number of element
+ * names. Instantiate the object with the number of elements that have values.
+ * Note that this can be increased after the fact but can (potentially) lead to
+ * memory fragmentation.
+ *
+ * For example:
+ *
+ * ```
+ *    enum class MyElementNames {
+ *      element0,
+ *      element1,
+ *      element2,
+ *      element3
+ *    };
+ *
+ *    using MyPackedStringStruct = PackedStruct<std::string, 4, MyElementNames>;
+ *
+ *    MyPackedStringStruct packed_struct(2); // For this instantiation, only 2
+ *                                           // elements are non-empty
+ *    packed_struct.set(MyElementNames::element1, "abc");
+ *    packed_struct.set(MyElementNames::element2, "def");
+ *
+ *    // Code that accesses the elements of the struct needs to check for existence first.
+ *    if (packed_struct.has(MyElementNames::element0) {
+ *      auto& element0 = packed_struct.get(MyElements::element0);
+ *    }
+ * ```
+ */
+template <class T, uint8_t max_size, class ElementName> class PackedStruct {
+public:
+  PackedStruct(size_t a_capacity = 0) : data_(a_capacity > 0 ? new T[a_capacity] : nullptr) {
```

Comment:
I'm guessing it would be awkward to set capacity correctly from the caller's perspective.

What would be the impact if we just always sized it the the number of elements? Sure it would be an n^2 initialization process but N would be quite small I think.

Reply:
Are you suggesting we always start with `capacity = max_size` and then introduce a method like `finalize` to resize it?

Reply:
Actually I was suggesting just starting capacity=0 and incrementing as fields are set.

Reply:
That makes more sense. I misunderstood, because that is already the default behaviour. Are you suggesting I get rid of the constructor argument entirely?

Reply:
I guess there's not really a compelling reason to remove it, but I'd love to understand how it would be used. Maybe you could add another Draft PR that patches this in for a config block, and makes use of it, as a demo?

Reply:
The main use case came from the now deleted macros, where I iterated over the proto message to see which fields had values first :-)

Reply:
oh this makes sense. does that functionality require macros or could we provide it via a templated API?

Reply:
I don't know how to do it without macros.

Reply:
SGTM. I'll add that when I use this class for memory optimisations.